File: leetcode/25.k-个一组翻转链表.ts
/*
 * @lc app=leetcode.cn id=25 lang=typescript
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseKGroup(head: ListNode | null, k: number): ListNode | null {
    const dummy = new ListNode(0, head);
    let pre = dummy;

    while (head) {
        let tail = pre;
        for (let i = 0; i < k; ++i) {
            tail = tail.next;
            if (!tail) return dummy.next;
        }

        const next = tail.next;
        [head, tail] = reverse(head, tail);
        pre.next = head;
        tail.next = next;
        pre = tail;
        head = tail.next;
    }

    return dummy.next;
};
function reverse(head: ListNode, tail: ListNode): [ListNode, ListNode] {
    let prev = tail.next;
    let cur = head;
    while (prev !== tail) {
        const next = cur.next;
        cur.next = prev;
        prev = cur;
        cur = next;
    }
    return [tail, head];
}
// @lc code=end

File: leetcode/25.k-个一组翻转链表.go
/*
 * @lc app=leetcode.cn id=25 lang=golang
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	dummy := &ListNode{0, head}
	node := dummy
	for head != nil {
		tail := node
		for i := 0; i < k; i++ {
			tail = tail.Next
			if tail == nil {
				return dummy.Next
			}
		}

		next := tail.Next
		head, tail = reverse(head, tail)
		tail.Next = next
		node.Next = head
		node = tail
		head = tail.Next
	}
	return dummy.Next
}
func reverse(head, tail *ListNode) (*ListNode, *ListNode) {
	prev := tail.Next
	node := head
	for prev != tail {
		next := node.Next
		node.Next = prev
		prev = node
		node = next
	}
	return tail, head
}
// @lc code=end

